src/ast/codegen_llvm.cpp

```diff
@@ -557,8 +557,16 @@ void CodegenLLVM::visit(Call &call)
       arg.offset = struct_layout->getElementOffset(i+1); // +1 for the printf_id field
     }
 
-    AllocaInst *printf_args = b_.CreateAllocaBPF(printf_struct, "printf_args");
-    b_.CreateMemSet(printf_args, b_.getInt8(0), struct_size, 1);
+    Value *printf_args = b_.CreateGetPrintfMap(printf_struct);
+
+    Function *parent = b_.GetInsertBlock()->getParent();
+    BasicBlock *zero = BasicBlock::Create(module_->getContext(), "printfzero", parent);
+    BasicBlock *notzero = BasicBlock::Create(module_->getContext(), "printfnotzero", parent);
+    b_.CreateCondBr(b_.CreateICmpNE(printf_args, ConstantExpr::getCast(Instruction::IntToPtr, b_.getInt64(0), PointerType::get(printf_struct, 0)), "printzerocond"), notzero, zero);
+
+    b_.SetInsertPoint(notzero);
+
+    b_.CreateProbeRead(printf_args, struct_size, ConstantExpr::getCast(Instruction::IntToPtr, b_.getInt64(reinterpret_cast<uintptr_t>(bpftrace_.print_map_zero_)), PointerType::get(printf_struct, 0)));
```

Comment:
`printf_args` is in BPF memory here, so wouldn't a regular `CreateMemSet()` work?

Reply:
memset will only work for small values (even if we're using a map). I'm on another branch right now, I'll try to share more information later and I'll add a comment here explaining why this is implemented this way.